File: model/ad_client.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type AdClient struct {
	ID           int            `json:"id"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `json:"deleted_at" gorm:"index" swaggertype:"string"`
	UserID       int            `json:"user_id"`
	User         *User          `json:"user"`
	Description  string         `json:"description"`
	Headline     string         `json:"headline"`
	Price        float64        `json:"price"`
	TypeID       int            `json:"type_id"`
	Type         Type           `json:"type"`
	CityID       int            `json:"city_id"`
	City         City           `json:"city"`
	Documents    []Document     `json:"documents" gorm:"many2many:ad_client_documents"`
	UrlDocuments []string       `json:"url_documents" gorm:"-"`
	StartDate    time.Time      `json:"start_date"`
	EndDate      Time           `json:"end_date"`
	Address      string         `json:"address"`
	Latitude     *float64       `json:"latitude"`
	Longitude    *float64       `json:"longitude"`
	Status       string         `json:"status"`
}

type FilterAdClient struct {
	UserID        *int
	Unscoped      *bool
	TypeID        *int
	SubCategoryID *int
	CityID        *int
	Deleted       *bool
	Limit         *int
	Offset        *int
	Status        *string
	ID            []int
	ASC           []string
	DESC          []string
}

type AdClientDocuments struct {
	AdClientID int
	DocID      int
}

type AdClientInteracted struct {
	ID         int            `json:"id"`
	CreatedAt  time.Time      `json:"created_at"`
	UpdatedAt  time.Time      `json:"updated_at"`
	DeletedAt  gorm.DeletedAt `json:"deleted_at" gorm:"index" swaggertype:"string"`
	AdClientID int            `json:"ad_client_id"`
	UserID     int            `json:"user_id"`
	UserRating float64        `json:"user_rating" gorm:"-"`
	User       User           `json:"user"`
}

File: mobile/lib/src/features/ad/data/models/ad_sm_models/ad_sm_client_model/ad_client.dart
import 'package:eqshare_mobile/src/core/data/models/category_models/sub_category.dart';
import 'package:eqshare_mobile/src/core/data/models/cities/city/city_model.dart';
import 'package:eqshare_mobile/src/core/data/models/document/document.dart';
import 'package:eqshare_mobile/src/core/data/models/user_model/user.dart';
import 'package:eqshare_mobile/src/features/ad/data/classes/ad_list_row_data.dart';
import 'package:eqshare_mobile/src/global_widgets/global_values.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'ad_client.freezed.dart';
part 'ad_client.g.dart';

@freezed
class AdClient with _$AdClient {
  const factory AdClient({
    required int id,
    String? description,
    required String headline,
    double? price,
    @JsonKey(name: 'type_id') int? typeId,
    SubCategory? type,
    List<Document>? documents,
    @JsonKey(name: 'start_date') String? startDate,
    @JsonKey(name: 'end_date') dynamic endDate,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
    @JsonKey(name: 'deleted_at') DateTime? deletedAt,
    @JsonKey(name: 'url_thumbnail') List<String>? urlThumbnail,
    User? user,
    required String address,
    double? latitude,
    double? longitude,
    String? status,
    City? city,
    @JsonKey(name: 'city_id') int? cityId,

  }) = _AdClient;
  factory AdClient.fromJson(Map<String, dynamic> json) =>
      _$AdClientFromJson(json);

  static AdListRowData getAdListRowDataFromSM(AdClient e,
      {bool? isClientType, String? imageUrl}) {
    String? imageUrl;

    if (e.documents == null || e.documents?.isEmpty == true) {
      imageUrl = imageUrl ?? '';

    } else {
      imageUrl = e.documents?.first.shareLink ?? imageUrl ?? '';

    }
    
    return AdListRowData(
        address: e.address,
        title: e.headline,
        allServiceTypeEnum: AllServiceTypeEnum.MACHINARY_CLIENT,
        deletedAt: e.deletedAt,
        id: e.id,
        price: e.price?.toDouble(),
        category: e.type?.name,
        city: e.city?.name,
        imageUrl: imageUrl,
        status: e.status,
        isClientType: isClientType,
        imageUrls: e.documents?.map((e) => e.shareLink ?? '').toList(),
        createdAt: e.createdAt.toString());
  }
}
